### DISCUSSION TITLE
Mojolicious logging with hypnotoad and systemd

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#1926] [Q&A]
Does anyone have good experience with mojo logging when running the applications via hypnotoad and systemd.

As far as I know the logging to stderr etc should be pucked up by systemd and is viewable in the journal.

As far as I can tell I can get my app to log something from the setup method, everything else just disappears.

Running hypnotoad manual in the foreground logs just fine just as we get with morbo.

I'm assuming that it is the logs from the workers that are lost?

Any ideas?

**Comment by kraih** [comment]
Just use prefork instead of Hypnotoad with systemd.

**Reply by daleif** [reply]
Don't we then loose the hot deployment feature? Which I kinda like.

Shouldn't this be mentioned in the hypnotoad section of https://docs.mojolicious.org/Mojolicious/Guides/Cookbook#Hypnotoad

**Reply by s1037989** [reply]
You won't lose hot deployment. See the section below referring to hot deployment using prefork and reuse=1

**Reply by daleif** [reply]
@s1037989 Interesting, why isn't the reuse thing a normal option. And it also sounds rather cumbersome, though one can probably build it into the systemd script

**Reply by Grinnz** [reply]
Reusable sockets are a newer option than hot deployment and require intentional setup - both the initial and new instance of the server have to be set up as reusable. It also doesn't work everywhere

**Reply by jhthorsen** [reply]
How do you set up systemd to not stop the previous before the new prefork server is up and running?

@daleif: STDERR and STDOUT is closed by hypnotoad, but you could use something like https://metacpan.org/pod/Mojolicious::Plugin::Syslog to send the log messages to syslog.

**Reply by s1037989** [reply]
@jhthorsen Does the Syslog plug-in work well as a deployment specific plugin? That is, could I have it set in the production conf but not the development conf, and when enabled with production it'll just redirect to Syslog all the $c->log calls that development would send to stdout/file?

**Reply by jhthorsen** [reply]
@s1037989: "enable" is disabled when environment is "development". (Look at the documentation for register()) Not sure if I understand the question though... You can load it or not load it whenever you want.

**Reply by daleif** [reply]
@jhthorsen `Plugin::Syslog` seems very interesting. I would suggest mentioning the STDERR and STDOUT closing by `hypnotoad` somewhere. Then others will not get confused.

The preforking works fine, but as far as I can see to restart one has to stop the service and start it again, which is rather annoying, compared to just doing a `systemctl reload`.

**Reply by jhthorsen** [reply]
@daleif: I guess it could be mentioned somewhere, but since Hypnotoad detaches itself from the terminal then there isn't really any sensible place to send STDERR / STDOUT. Maybe create a documentation PR? :)

I'm also curious (as mentioned earlier) how "systemctl reload" would work with multiple multiple processes using reuse=1.

**Reply by daleif** [reply]
I'm wondering, what happens if one explicitly instruct the app to use a specific file for logging. Will the workers under hypnotoad then be able to log to it?

**Reply by daleif** [reply]
Using a specific file to log to seems to be working as well, at least I'm getting logs in from the worker processes. Just remember the user specified under systemd config needs to have write access to the folder with the `log` file.

Just need a proper place for the logs which can also be configured with logrotate or similar.

**Reply by guest20** [reply]
I think you just want to run `hypnotoad -f` under `systemd`, that way it doesn't close the handles and `journald` can slurp up the logs from STDOUT/ERR

**Reply by daleif** [reply]
@guest20 to be honest, I kinda like having the logs of each app in a separate log file

**Comment by daleif** [comment]
In the end I've opted to to have startup set a specific log file when in production

```
  if ( $self->mode eq 'production' && $config->{'log_file_production'} ) {
      $self->log( Mojo::Log->new( path => $config->{'log_file_production'},
      				  level => 'info') );
    }
```
seems to cover all my needs.

One additional question. When combining this with `log_rotate`, does the `hypnotoad` process need to be reloaded? (I noticed that reloading systemd services was one of the things that `log_rotate` could do).

**Reply by s1037989** [reply]
I know this wasn't your question, but you can use [deployment-specific configuration files](https://docs.mojolicious.org/Mojolicious/Plugin/Config#DESCRIPTION), e.g. app.development.conf and app.production.conf.  Each would specify the log file to use, and even whether or not to use a log file or just log to stdout.

```
sub startup ($self) {
  my $cfg = $self->plugin('Config');
  $self->log->path($cfg->{log_file}) if $cfg->{log_file};
}
```

**Reply by daleif** [reply]
Interesting. Did not know that. 

Am I reading it correctly that `$moniker.$mode.conf` will extend/overwrite what is in `$moniker.conf`

That would be useful. Thank you for the hint.

**Comment by jhthorsen** [comment]
If you want, you could try out https://metacpan.org/pod/Mojo::Server::DaemonControl
